test/numbers.jl

```diff
@@ -2790,9 +2790,13 @@ let types = (Base.BitInteger_types..., BigInt, Bool,
                 # broken, fixme then remove this branch
                 @test_throws ErrorException @inferred(Base.promote_op(op, R, S))
                 T = Base.promote_op(op, R, S)
-                if ((R === Bool || S === Bool) && op in (+, *)) ||
-                       ((S in (Rational{Int}, Complex{Float16})) && op === ^) ||
-                       (R === Complex{Float16} && op === ^)
+                if (Bool in (R, S) && op in (+, *)) ||
+                        (Complex{Float16} in (R, S) && op === ^) ||
+                        (R in (Int, Rational{Int}, Complex{Int})) ||
+                        (S in (Int, Rational{Int})) ||
+                        (S === Complex{Int} && op in (+, -, *, ^)) ||
+                        (Complex{UInt} in (R, S) && op === ^) ||
+                        (S <: Integer && op === ^)
```

Comment:
This is really messy and took more time to get passing than I should have spent on it. We should maybe comment out `Float16` and `Complex{Float16}` from the loop if this is hard to fix, though we won't know when it gets fixed if we do that.